File: 2015/20/rust/src/main.rs
// --- Constants --
const INPUT_FILE: &'static str = "../input/input.txt";

// --- Main function ---
fn main() -> Result<(), Box<dyn std::error::Error>> {
    // --- Read and parse the input file ---
    let input = std::fs::read_to_string(INPUT_FILE)?.trim().parse::<usize>()?;

    // --- Puzzle 1 ---
    let mut house = 0;
    let mut presents = 0;
    while presents < input {
        house += 1;
        presents = 10 * divisors::get_divisors(house).iter().sum::<usize>();
        // --- divisors doen't take into account 1 as divisor or the number itself ---
        if house <= 2 {
            presents += 10;
        } else {
            presents += 10 * (house + 1);
        }
    }
    println!("1. House number: {}", house);

    // --- Puzzle 2 ---
    let mut house = 0;
    let mut presents = 0;
    while presents < input {
        house += 1;
        presents = 11 * divisors::get_divisors(house).iter()
            .filter(|x| 50 * **x >= house)
            .sum::<usize>();
        // --- divisors doen't take into account 1 as divisor or the number itself ---
        if house <= 2 {
            presents += 11;
        } else {
            if house <= 50 { presents += 11; }
            presents += 11 * house;
        }
    }
    println!("2. House number: {}", house);

    Ok(())
}

File: 2015/20/go/main.go
package main

import (
	"fmt"
	"log"
	"math"
	"os"
	"strconv"
	"strings"
)

// --- Constants ---
const FILENAME = "../input/input.txt"

// --- Functions ---
func divisorsSum(n int) int {
	sum := 1
	for i := 2; i*i <= n; i++ {
		if n%i == 0 {
			sum += i
			if i*i != n {
				sum += n / i
			}
		}
	}
	return sum
}

func divisors(n int) []int {
	var a []int
	if n > 1 {
		a = append(a, 1, n)
	} else {
		a = append(a, 1)
	}

	sqrtN := int(math.Sqrt(float64(n)))
	for i := 2; i <= sqrtN; i++ {
		if n%i == 0 {
			a = append(a, i)
			if i*i != n {
				a = append(a, n/i)
			}
		}
	}
	return a
}

func presentsForHouse(house int) int {
	divs := divisors(house)
	var filtered []int
	for _, d := range divs {
		if d*50 >= house {
			filtered = append(filtered, d)
		}
	}
	sum := 0
	for _, x := range filtered {
		sum += x
	}
	return 11 * sum
}

func main() {
	// --- Variables ---
	house := 0
	presents := 0

	// --- Read input file ---
	bytes, err := os.ReadFile(FILENAME)
	if err != nil {
		log.Fatal(err)
	}
	text := strings.TrimSpace(string(bytes))
	input, err := strconv.Atoi(text)
	if err != nil {
		log.Panic(err)
	}

	// --- Puzzle 1 ---
	for presents < input {
		house++
		presents = 10 * divisorsSum(house)
		if house <= 2 {
			presents += 10
		} else {
			presents += 10 * (house + 1)
		}
	}
	fmt.Println("1. House number:", house)

	// --- Puzzle 2 ---
	house = 0
	presents = 0
	for presents < input {
		house++
		presents = presentsForHouse(house)
	}
	fmt.Println("2. House number:", house)
}
